### PR TITLE
7268 Checking vmem limit of wrong process for door_calls

### PR DESCRIPTION
**Pull request by prakashsurya**
Reviewed by: Dan Kimmel <dan.kimmel@delphix.com>
Reviewed by: George Wilson <george.wilson@delphix.com>
Reviewed by: Sebastien Roy <sebastien.roy@delphix.com>

As part for the Solaris door_call, the client/caller thread specifies a
buffer to hold the results. If the buffer is not big enough to hold the
results, the server/callee thread creates a new buffer in the caller's
address space. See the door_call manual page for more specifics.

The problem we discovered was when the caller (in our case a 64-bit JVM
with more than 4G vmem allocated) made a door_call to a callee (32-bit
ipmgmtd); a check was made against the virtual memory size limit of the
callee even though the memory is being allocated in the caller's address
space. This triggered an error which effectively prevents our 64-bit JVM
process from using more than 3GB for the Java heap. Although in this
case, the issue happened between 64-bit and 32-bit processes, I believe
it would also occur if both are 32-bit, but the callee has a lower
"ulimit -v".

It appears this can be fixed by using the process that owns the address
space we are allocating memory in, as->a_proc, instead of the current
process. All p is used for in as_map_loced is process limitation
checks (p_vmem_ctl and p_rctls) so it appears safe and correct to
instead use the address space's process. There were no other references
to curproc in this file.

When creating a shared memory segment a_proc will be NULL, so we still
use curproc in that case.

Upstream Bugs: DLPX-42516

### CODE REVIEW

**Comment by rmustacc** [comment]
I don't think this change is quite safe as is. The problem is that the as structures come from a kmem cache. That means that as long as objects aren't constructed or destructed their members are going to be recycled. Importantly, in the many members that get cleared after allocation in `as_alloc()`, the `a_proc` member used here is not reset.

While many of the address space routines will go and reset the value of `a_proc` themselves. There are some that don't, like in `shmem_lock()`. In addition the path that frees these back to its cache, `as_free()` does not reset this member. As the majority of the address spaces that show up are going to have processes so the freed objects will have them. This means that it is possible for the callers like `shmem_lock()` to get an old value of `a_proc` set, which it both won't clean up and will be invalid.

While it's pretty rare this kind of thing has bitten us in the past (see https://www.illumos.org/issues/3672 for an example). Thankfully the solution here is pretty simple we need to set the a_proc member to NULL in `as_alloc()`.

**Comment by prakashsurya** [comment]
Good catch! Yes, I've been bitten by that sort of issue in the past too. I'll update `as_alloc`; assuming that's safe to do, as you suggested.

File: usr/src/uts/common/vm/vm_as.c [hunk 1]
```diff
@@ -665,6 +666,7 @@ as_alloc(void)
 	as->a_lastgap		= NULL;
 	as->a_lastgaphl		= NULL;
 	as->a_callbacks		= NULL;
+	as->a_proc		= NULL;
```
Comment (prakashsurya) [inline usr/src/uts/common/vm/vm_as.c]: @rmustacc I added this line per your prior comment. Please let me know if this isn't what you had in mind.

Comment (rmustacc) [inline usr/src/uts/common/vm/vm_as.c]: Yes, I think that's right.

**Comment by prakashsurya** [comment]
Failure in `zvol/zvol_swap/setup` is a known issue.

@rmustacc can I count you has having reviewed and approved this?

**Comment by rmustacc** [comment]
@prakashsurya You can count me as a reviewer.

**Pull request closed without merging**